File: src/services/verification_service.py
from uuid import UUID
from datetime import datetime, timezone

from fastapi import HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from src.config import get_settings
from src.models.schemas.responses import VerificationResponse
from src.models.database.certificate import Certificate
from src.core.crypto.hmac_handler import HMACHandler
from src.core.crypto.token_manager import TokenManager
from src.core.cache.redis_manager import RedisManager


class VerificationService:
    def __init__(
        self,
        hmac_handler: HMACHandler,
        token_manager: TokenManager,
        redis_manager: RedisManager,
    ):
        self.hmac = hmac_handler
        self.token_manager = token_manager
        self.redis = redis_manager

    async def verify_certificate(
        self, certificate_id: UUID, signature: str, db: AsyncSession
    ) -> VerificationResponse:
        """Main verification flow"""

        # Get certificate with cache check
        cached = await self.redis.get_certificate_status(str(certificate_id))

        if cached:
            is_valid = cached["is_valid"]
            expires_at = datetime.fromisoformat(cached["expires_at"])
            cert = None  # type: ignore
        else:
            # Query database
            stmt = select(Certificate).where(Certificate.id == certificate_id)
            result = await db.execute(stmt)
            cert: Certificate = result.scalar_one_or_none()  # type: ignore

            if not cert:
                raise HTTPException(status_code=404, detail="Certificate not found")

            expires_at = cert.expires_at
            is_valid = not cert.revoked_at and expires_at > datetime.now(timezone.utc)

            await self.redis.cache_certificate_status(
                str(certificate_id), is_valid, expires_at
            )

        if not is_valid:
            raise HTTPException(
                status_code=410, detail="Certificate has expired or been revoked"
            )

        if not cert:
            # Re-fetch if we only had cached data
            stmt = select(Certificate).where(Certificate.id == certificate_id)
            result = await db.execute(stmt)
            cert: Certificate = result.scalar_one()

        if not self.hmac.verify_signature(certificate_id, cert.qr_nonce, signature):  # type: ignore
            raise HTTPException(status_code=401, detail="Invalid signature")

        session_token, ttl = self.token_manager.create_session_token(
            certificate_id, expires_at
        )

        return VerificationResponse(
            session_token=session_token,
            expires_in_seconds=ttl,
            file_endpoint=f"/api/certificates/{certificate_id}/content",
            certificate_title=cert.certificate_title,
            holder_name=cert.holder_name,
            expires_at=expires_at,
            is_valid=is_valid,
        )


settings = get_settings()


def get_verification_service():
    hmac_handler = HMACHandler(settings.secret_key)
    token_handler = TokenManager(settings.secret_key)
    redis_manager = RedisManager(settings.redis_url, settings.redis_cache_ttl_seconds)
    return VerificationService(hmac_handler, token_handler, redis_manager)

File: app/Services/VerificationService.php
<?php

namespace App\Services;

use App\Models\Certificate;
use Illuminate\Support\Facades\Cache;
use Illuminate\Support\Facades\Crypt;
use Illuminate\Support\Str;
use Carbon\Carbon;
use Symfony\Component\HttpKernel\Exception\HttpException;

class VerificationService
{
    protected string $secretKey;
    protected int $cacheTtl;

    public function __construct()
    {
        $this->secretKey = config('app.key');
        $this->cacheTtl = config('cache.certificate_ttl', 3600);
    }

    /**
     * Executes the primary cryptographic validation flow for an incoming scan.
     *
     * @throws HttpException
     */
    public function verifyCertificate(string $certificateId, string $signature): array
    {
        $cacheKey = "certificate:full:{$certificateId}";

        // 1. Fetch the certificate. Instead of querying twice if the status is cached,
        // we cache the full model payload to dramatically reduce DB IO operations.
        $cert = Cache::remember($cacheKey, $this->cacheTtl, function () use ($certificateId) {
            return Certificate::find($certificateId);
        });

        if (!$cert) {
            throw new HttpException(404, "Certificate not found");
        }

        // 2. Validate structural integrity state boundaries
        $now = Carbon::now('UTC');
        $isValid = ($cert->revoked_at === null) && ($cert->expires_at->isAfter($now));

        if (!$isValid) {
            throw new HttpException(410, "Certificate has expired or been revoked");
        }

        // 3. Cryptographic Signature Authentication check
        if (!$this->verifySignature($certificateId, $cert->qr_nonce, $signature)) {
            throw new HttpException(401, "Invalid signature");
        }

        // 4. Issue a secure, short-lived session token
        [$sessionToken, $ttlSeconds] = $this->createSessionToken($certificateId, $cert->expires_at);

        return [
            'session_token'       => $sessionToken,
            'expires_in_seconds'  => $ttlSeconds,
            'file_endpoint'       => "/api/v1/certificates/{$certificateId}/content",
            'certificate_title'   => $cert->certificate_title,
            'holder_name'         => $cert->holder_name,
            'expires_at'          => $cert->expires_at->toIso8601String(),
            'is_valid'            => true,
        ];
    }

    /**
     * Verify that the signature computed from the QR matches our application key.
     */
    protected function verifySignature(string $certificateId, string $nonce, string $signature): bool
    {
        $payload = "{$certificateId}:{$nonce}";
        $computedSignature = hash_hmac('sha256', $payload, $this->secretKey);

        return hash_equals($computedSignature, $signature);
    }

    /**
     * Create an encrypted session token tied to the certificate lifespan.
     */
    protected function createSessionToken(string $certificateId, Carbon $expiresAt): array
    {
        $now = Carbon::now('UTC');
        
        // Dynamic TTL: Caps token lifespan to the exact time the certificate expires,
        // or a default fallback threshold (e.g., 15 minutes), whichever is shorter.
        $maxTtlSeconds = 900; 
        $remainingCertLife = $now->diffInSeconds($expiresAt, false);
        $ttlSeconds = min($maxTtlSeconds, max(0, $remainingCertLife));

        // Package payload data tightly
        $tokenPayload = [
            'cert_id'    => $certificateId,
            'jti'        => Str::random(16),
            'expires_at' => $now->addSeconds($ttlSeconds)->timestamp
        ];

        // Encrypt symmetrically using AES-256-CBC via Laravel's native system key
        $sessionToken = Crypt::encrypt($tokenPayload);

        return [$sessionToken, $ttlSeconds];
    }
}